scanpy/get/_aggregated.py

```diff
@@ -0,0 +1,437 @@
+        )
+        if normalize:
+            n_row = A.shape[0]
+            row_sums = np.asarray(A.sum(axis=1))
+            D = dia_matrix(((row_sums.T**-1), [0]), shape=(n_row, n_row))
+            A = D * A
+        return A, keys
+
+    def _extract_indices(self):
+        def _filter_indices(key_set, keys, key_index, df_index, weight_value=None):
+            keep = [i for i, k in enumerate(keys) if k in set(key_set)]
+            if len(keep) == 0:
+                raise ValueError("No keys in key_set found in adata.obs[key].")
+            elif len(keep) < len(keys):
+                mask = np.in1d(key_index, keep)
+                remap = np.zeros(len(keys), dtype=np.int64)
+                for i, j in enumerate(keep):
+                    remap[j] = i
+                keys = [keys[j] for j in keep]
+                key_index = np.array(
+                    [remap[i] for i in key_index[mask]], dtype=np.int64
+                )
+                df_index = df_index[mask]
+                if weight_value is not None:
+                    weight_value = weight_value[mask]
+            return keys, key_index, df_index, weight_value
+
+        key_value = self._groupby
+        keys, key_index = np.unique(_ndarray_from_seq(key_value), return_inverse=True)
+        df_index = np.arange(len(key_index))
+        if self._weight is None:
+            weight_value = None
+        else:
+            weight_value = self._weight.values[df_index]
+        if self._key_set is not None:
+            keys, key_index, df_index, weight_value = _filter_indices(
+                self._key_set, keys, key_index, df_index, weight_value
+            )
+        self._key_index = key_index  # passed to count and count_mean_var to avoid re-extracting in the latter
+        return keys, key_index, df_index, weight_value
+
+
+def _power(X, power):
+    return X ** power if isinstance(X, np.ndarray) else X.power(power)
+
+
+def _ndarray_from_seq(lst: Sequence):
+    # prevents expansion of iterables as axis
+    n = len(lst)
+    if n > 0 and isinstance(lst[0], cabc.Iterable):
+        arr = np.empty(n, dtype=object)
+        arr[:] = lst
+    else:
+        arr = np.array(lst)
+    return arr
+
+
+def _superset_columns(df: pd.DataFrame, groupby_key: str) -> List[str]:
+    """\
+    Find all columns which are a superset of the key column.
+
+    Params
+    ------
+    df
+        DataFrame which contains candidate columns.
+    groupby_key
+        Key for column of which to find superset of columns.
+
+    Returns
+    -------
+    Superset columns.
+    """
+    columns = []
+    groupy_key_codes = df[groupby_key].astype('category')
+    for key in df:
+        if key != groupby_key:
+            key_codes = df[key].astype('category')
+            if all(
+                [
+                    key_codes[groupy_key_codes == group_key_code].nunique() == 1
+                    for group_key_code in groupy_key_codes
+                ]
```

Comment:
Don’t create an unnecessary list here. If you leave off the brackets, you pass a generator object to the function that lazily calculates values one by one. Since `all` exits early when it encounters `False`, the following code therefore does less work when that happens (and is formatted prettier).

```suggestion
                key_codes[groupy_key_codes == group_key_code].nunique() == 1
                for group_key_code in groupy_key_codes
```